src/Symfony/Component/Clock/HrClock.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Clock;
+
+/**
+ * A high-resolution monotonic clock, counting from an arbitrary point in time.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class HrClock implements TimeInterface
+{
+    private \DateTimeZone $timezone;
+
+    public function __construct(\DateTimeZone|string $timezone = null)
+    {
+        if (false === hrtime()) {
+            throw new \LogicException('hrtime() returned false: the runtime environment does not provide access to high-resolution time.');
+        }
+
+        if (\is_string($timezone ??= date_default_timezone_get())) {
+            $this->timezone = new \DateTimeZone($timezone);
+        } else {
+            $this->timezone = $timezone;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function now(): \DateTimeImmutable
+    {
+        $hrtime = hrtime();
+        $now = '@'.$hrtime[0].'.'.substr(str_pad($hrtime[1], 9, 0, \STR_PAD_LEFT), 0, -3);
+
+        return (new \DateTimeImmutable($now, $this->timezone))->setTimezone($this->timezone);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function timeArray(): array
```

Comment:
Not a fan of using arrays, less so in interfaces.

Besides, I'm not sure why this interface needs anything beyond `now`. Everything else looks like it doesn't belong here. This class is mixing too many responsibilities. I think there should be a difference between a Clock interface, whose job is merely to give the current time, and a component for all things time-related.

It's better to start small and add/compose only if absolutely needed. Splitting an interface later will be a nightmare.

Reply:
Hi Anna, thanks for chiming in.

> Not a fan of using arrays, less so in interfaces.

Yep, yet it's the common way to return a tuple of numbers when the size of integers is too small to properly count the time, typically on 32-bit arch. I don't think adding a VO for that is worth it.

>  I'm not sure why this interface needs anything beyond now

The plan is to have PSR-20 be the super simple interface that fits most use cases.

`TimeInterface` will extend PSR-20 once it's available and will be useful when performance matters (using DateTimeImmutable creates overhead that isn't desired when eg writting a StopWatch) or when needing to abstract sleep().

Of course, until PSR-20 is available, ppl won't have a better choice than using `TimeInterface`. I considered adding a `ClockInterface` that contain only `now()` but I'd better not since PSR-20 might just be the one. IF it gets released soon enough of course. I'm fine adding `ClockInterface` here also but we're going to have a hard time deprecating it when PSR-20 will come. I'd rather encourage the PHP-FIG to release PSR-20 by November...

Reply:
As already mentioned: there is an interface that is used by the most-used Clock-Libraries that is also forward-compatible to the current draft of PSR-20 that could be used *right away*

Reply:
> As already mentioned: there is an interface that is used by the most-used Clock-Libraries that is also forward-compatible to the current draft of PSR-20 that could be used right away

Sorry I didn't comment on this topic: we never referenced any "interop" packages as that doesn't fit out maintenance policies, per the package principles. We can already anticipate that referencing [stella-maris/clock](https://packagist.org/packages/stella-maris/clock) would come at a price that the community will have to pay when we'll replace it with PSR-20. Instead, I prefer have the community use `TimeInterface` even if it does a bit too much for the common use case: this won't incur any "forced" migration cost and will be easy to migrate to PSR-20 when possible.

Reply:
The `stella-maris/clock` package will extend `psr/clock` once it is released (and based on the current draft). So all packages implementing or extending that interface will immediately implement or extend `psr/clock`. And it will stay as long as necessary so that the community will not have to pay any price at all when it will be replaced by PSR-20. Provided that PSR-20 will be released as it is currently drafted.

Just as a background-information.

Reply:
I think that this is suffering from scope creep. The initial problem is "how do we unit-test code that relies on the current date?" A simple interface with `now` solves it quickly and painlessly, in a dozen lines of code. The need to conform to some other interface is unnecessary in this context. You can always create a bigger component later (if you have a problem that needs solving), in which you can then reuse the `DateTimeClock`. You don't need to figure out the entire roadmap of time handling to solve this one small problem.

This is from my blog. That's all you really need:
```php
interface Clock
{
    public function now(): DateTimeImmutable;
}

final class DateTimeClock implements Clock
{
    public function now(): DateTimeImmutable
    {
        return new DateTimeImmutable(
            'now', 
            new DateTimeZone('UTC')
        );
    }
}
```